### PR TITLE
emacs-lisp: Add bindings to lisp-interaction-mode and new eval functions

### PR DESCRIPTION
**Pull request by justbur**
Also declare prefixes

### CODE REVIEW

**Comment by justbur** [comment]
possible start to solution for  #3103

**Comment by robbyoconnor** [comment]
:+1:

File: layers/+lang/emacs-lisp/extensions.el [hunk 1]
```diff
@@ -15,17 +15,59 @@
...
+      "mec" 'spacemacs/eval-current-form
+      "mee" 'eval-last-sexp
+      "mer" 'eval-region
+      "mef" 'eval-defun
+      "mel" 'lisp-state-eval-sexp-end-of-line
+      "m,"  'lisp-state-toggle-lisp-state
+      "mtb" 'spacemacs/ert-run-tests-buffer
+      "mtq" 'ert))
+
+  (if (configuration-layer/layer-usedp 'spacemacs)
```
Comment (syl20bnr) [inline layers/+lang/emacs-lisp/extensions.el]: Do it the other way around, add `emacs-builtin-emacs-lisp` to the layer `spacemacs` and di your stuff in a post-init.

Comment (justbur) [inline layers/+lang/emacs-lisp/extensions.el]: I don't quite understand. What if someone had Spacemacs-base and emacs-lisp
layers?

Comment (syl20bnr) [inline layers/+lang/emacs-lisp/extensions.el]: Then put it in `spacemacs-base`, I said `spacemacs` because this is what you test.

Comment (justbur) [inline layers/+lang/emacs-lisp/extensions.el]: I need to make sure smartparens is available, which is currently not in -base. Sorry, I'm not up to date on how all of the init/extentions/built-in package stuff works at the moment, which was part of the reason I asked.

Comment (TheBB) [inline layers/+lang/emacs-lisp/extensions.el]: If it depends on smartparens, it should be in a `post-init-smartparens` function.

Comment (justbur) [inline layers/+lang/emacs-lisp/extensions.el]: @TheBB that's right, thanks.

**Comment by justbur** [comment]
@syl20bnr @TheBB how does it look now?

**Comment by TheBB** [comment]
Yeah, now it looks reasonable.

**Comment by justbur** [comment]
great, squashed

File: layers/+lang/emacs-lisp/packages.el [hunk 2]
```diff
@@ -98,9 +103,47 @@
...
+      (evil-leader/set-key-for-mode mode
+        "m=b" 'srefactor-lisp-format-buffer
+        "m=d" 'srefactor-lisp-format-defun
+        "m=o" 'srefactor-lisp-one-line
+        "m=s" 'srefactor-lisp-format-sexp))))
+
+(defun emacs-lisp/post-init-smartparens ()
+  (if (version< emacs-version "25")
+      (ad-disable-advice 'preceding-sexp 'around 'evil)
+    (advice-remove 'elisp--preceding-sexp 'evil--preceding-sexp))
```
Comment (syl20bnr) [inline layers/+lang/emacs-lisp/packages.el]: I like the add-functions and advice API, I think we will migrate all advices to this API in 0.105.0, I'm ok to drop 24.3 at this point, 25 should be released soon.

Comment (syl20bnr) [inline layers/+lang/emacs-lisp/packages.el]: To be clear, it was just a remark, no need to change anything.

File: layers/+lang/emacs-lisp/extensions.el [hunk 3]
```diff
@@ -15,17 +15,24 @@
...
+      "meb" 'eval-buffer
+      "mee" 'eval-last-sexp
+      "mer" 'eval-region
+      "mef" 'eval-defun
+      "mel" 'lisp-state-eval-sexp-end-of-line
+      "m,"  'lisp-state-toggle-lisp-state
+      "mtb" 'spacemacs/ert-run-tests-buffer
+      "mtq" 'ert))
+
+  (unless (configuration-layer/package-usedp 'smartparens)
```
Comment (syl20bnr) [inline layers/+lang/emacs-lisp/extensions.el]: I would move this code to smartparens post-init

Comment (TheBB) [inline layers/+lang/emacs-lisp/extensions.el]: But it's a negative condition... how would you do that?

Comment (syl20bnr) [inline layers/+lang/emacs-lisp/extensions.el]: Oops, forget what I said :-)

File: layers/+lang/emacs-lisp/packages.el (hunk 2 above)
Comment (justbur) [inline layers/+lang/emacs-lisp/packages.el]: I'm having trouble figuring out when the new functions came into effect. Is it 24.4?

Comment (syl20bnr) [inline layers/+lang/emacs-lisp/packages.el]: Yes this is 24.4

Comment (justbur) [inline layers/+lang/emacs-lisp/packages.el]: ok, let me change the test then

**Comment by syl20bnr** [comment]
Thank you again (like everyday :-D)
Cherry-picked into develop branch, you can safely delete your branch.

**Pull request closed without merging**